### ISSUE TITLE
staticcheck: lint:ignore directives don't work inside multi-line function calls

### ISSUE THREAD

**Issue by cespare** [dominikh/go-tools#562]
This is staticcheck 2019.2.2, but I believe it's a longstanding issue.

If you put a lint directive within a multi-line function call, it will not work correctly:

``` go
func F() error {
	return fmt.Errorf(
		//lint:ignore ST1005 Proper noun; hard to rephrase.
		"Postgres blah blah blah %d %d %d",
		1, 2, 3,
	)
}
```

```
linty.go:6:19: error strings should not be capitalized (ST1005)
linty.go:7:3: this linter directive didn't match anything; should it be removed? ()
```

(I just noticed those empty parens in the second message. Should they be removed?)

You have to put the lint directive before the function call:

```
func F() error {
	//lint:ignore ST1005 Proper noun; hard to rephrase.
	return fmt.Errorf(
		"Postgres blah blah blah %d %d %d",
		1, 2, 3,
	)
}
```

This isn't all that surprising, to be honest, and also it may be very well be difficult to fix depending on what the AST gives you. I'll understand if the resolution is "unfortunate".

The [lint:ignore](http://staticcheck.io/docs/#line-based-linter-directives) docs just say "the following line of code", so maybe that should be expanded if the behavior can't be changed.

**Comment by dominikh** [comment]
I admit that this might be slightly confusing, but it is actually working as documented. We're flagging the function call, not the first argument of the function. As such, the linter directive has to precede the function call. This is reflected in the position information in the error message, too – note how we complain about capitalized errors on line 6, and about the ineffective linter directive on line 7 - the function argument would be on line 8.

One could argue whether the call or the argument should be flagged by staticcheck. In some places we make a conscious decision on which to flag. For example, ST1005 flags the function call because we're not happy with the error value that is being constructed – the fact that the error is based on an argument is coincidental. SA9002 (which complains about non-octal os.FileMode's) on the other hand flags function arguments because it complains about specific integer literals.

This might seem somewhat arbitrary, but at least it can be reasoned about, and I feel fairly confident about the explicit decisions I have made. 

What's more problematic are instances where we flag calls instead of arguments simply because we don't have the necessary position information – All checks that are based on a particular internal framework (`call rules`) lack position information for arguments due to its use of SSA form, and thus always flag the function call containing the argument instead. This is for example the case for `regexp.MustCompile("+")`. This wasn't always the case, though. We used to make the effort to obtain the missing position information, but for some reason we stopped doing that: https://github.com/dominikh/go-tools/blob/fc9f610700dc9b2e970f97648006947671b1e06b/staticcheck/lint.go#L2858 

We should probably fix that. Changing what we flag does have the risk of invalidating existing linter directives. However, we evidently already changed behavior once, and nobody complained. Most function calls are probably one-liners.

> (I just noticed those empty parens in the second message. Should they be removed?)

They shouldn't be empty :) Will fix.

tl;dr: problems do exist, but this issue didn't hit one of them. I'll see about fixing them.
